python/lsst/ip/diffim/imageDecorrelation.py

```diff
@@ -52,188 +54,211 @@ def computeClippedImageStats(im):
+
+        @param kappa  A matching kernel array derived from Alard & Lupton PSF matching
+        @param im2_psf The uncorrected psf array of the science image (and also of the diffim)
+        @param sig1   Average sqrt(variance) of template image used for PSF matching
+        @param sig2   Average sqrt(variance) of science image used for PSF matching
+        @return a 2-d numpy.array containing the new PSF
+        """
+        def post_conv_psf_ft2(psf, kernel, sig1=1., sig2=1.):
+            # Pad psf or kernel symmetrically to make them the same size!
+            # Note this assumes they are both square (width == height)
+            if psf.shape[0] < kernel.shape[0]:
+                diff = (kernel.shape[0] - psf.shape[0]) // 2
+                psf = np.pad(psf, (diff, diff), mode='constant')
+            elif psf.shape[0] > kernel.shape[0]:
+                diff = (psf.shape[0] - kernel.shape[0]) // 2
+                kernel = np.pad(kernel, (diff, diff), mode='constant')
+            psf_ft = fft2(psf)
+            kft = fft2(kernel)
+            out = psf_ft * np.sqrt((sig1**2 + sig2**2) / (sig1**2 + sig2**2 * kft**2))
+            return out
+
+        def post_conv_psf(psf, kernel, sig1=1., sig2=1.):
+            kft = post_conv_psf_ft2(psf, kernel, sig1, sig2)
+            out = ifft2(kft)
+            return out
+
+        pcf = post_conv_psf(psf=im2_psf, kernel=kappa, sig1=sig2, sig2=sig1)
+        pcf = pcf.real / pcf.real.sum()
+        return pcf
+
+    def fixEvenKernel(self, kernel):
+        """! Take a kernel with even dimensions and make them odd, centered correctly.
+        @param kernel a numpy.array
+        @return a fixed kernel numpy.array
+        """
+        # Make sure the peak (close to a delta-function) is in the center!
+        maxloc = np.unravel_index(np.argmax(kernel), kernel.shape)
+        out = np.roll(kernel, kernel.shape[0]//2 - maxloc[0], axis=0)
+        out = np.roll(out, out.shape[1]//2 - maxloc[1], axis=1)
+        # Make sure it is odd-dimensioned by trimming it.
+        if (out.shape[0] % 2) == 0:
+            maxloc = np.unravel_index(np.argmax(out), out.shape)
+            if out.shape[0] - maxloc[0] > maxloc[0]:
+                out = out[:-1, :]
+            else:
+                out = out[1:, :]
+            if out.shape[1] - maxloc[1] > maxloc[1]:
+                out = out[:, :-1]
+            else:
+                out = out[:, 1:]
         return out
 
-    pcf = post_conv_psf(psf=im2_psf, kernel=kappa, sig1=sig2, sig2=sig1)
-    pcf = pcf.real / pcf.real.sum()
-    return pcf
+    def doConvolve(self, exposure, kernel, use_scipy=False):
+        """! Convolve an Exposure with a convolution kernel.
+        @param exposure Input afw.image.Exposure to be convolved.
+        @param kernel Input 2-d numpy.array to convolve the image with
+        @param use_scipy Use scipy to do convolution instead of afwMath
+        @return a new Exposure with the convolved pixels.
```

Comment:
This appears to be incorrect, since you return two items: the exposure and `kern`.

Why return the convolved exposure, since you apparently modify it in place? Best to do one or the other.

For any arguments that are modified please mark them @param[in,out]